polygeist/lib/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.cpp

```diff
@@ -0,0 +1,230 @@
+//===- ReachingDefinitionAnalysis.cpp - Reaching Defs Analysis ------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Polygeist/Analysis/ReachingDefinitionAnalysis.h"
+#include "mlir/Analysis/AliasAnalysis.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "reaching-definition-analysis"
+
+namespace mlir {
+namespace polygeist {
+
+//===----------------------------------------------------------------------===//
+// ReachingDefinition
+//===----------------------------------------------------------------------===//
+
+raw_ostream &operator<<(raw_ostream &os, const ReachingDefinition &lastDef) {
+  if (lastDef.valueToModifiers.empty() &&
+      lastDef.valueToPotentialModifiers.empty())
+    return os.indent(4) << "<empty>\n";
+
+  auto printMap =
+      [&os](const DenseMap<Value, ReachingDefinition::ModifiersTy> &map,
+            StringRef title) {
+        for (const auto &entry : map) {
+          Value val = entry.first;
+          const ReachingDefinition::ModifiersTy &valModifiers = entry.second;
+
+          os.indent(4) << val << "\n";
+          os.indent(4) << title << "\n";
+          if (valModifiers.empty())
+            os.indent(6) << "<none>\n";
+          else {
+            for (Operation *op : valModifiers)
+              os.indent(6) << *op << "\n";
+          }
+          os << "\n";
+        }
+      };
+
+  printMap(lastDef.valueToModifiers, "mods:");
+  printMap(lastDef.valueToPotentialModifiers, "pMods:");
+  return os;
+}
+
+ChangeResult ReachingDefinition::join(const AbstractDenseLattice &lattice) {
+  ChangeResult result = ChangeResult::NoChange;
+  auto otherReachingDef = static_cast<const ReachingDefinition &>(lattice);
+
+  auto join = [&result](DenseMap<Value, ModifiersTy> &currentMap,
+                        DenseMap<Value, ModifiersTy> &otherMap) {
+    for (const auto &entry : otherMap) {
+      Value val = entry.first;
+      const ModifiersTy &newModifiers = entry.second;
+
+      ModifiersTy &currentModifiers = currentMap[val];
+      if (currentModifiers != newModifiers) {
+        currentModifiers.insert(newModifiers.begin(), newModifiers.end());
+        result |= ChangeResult::Change;
+      }
```

Comment:
```suggestion
      const std::size_t size = currentModifiers.size();
      currentModifiers.insert(newModifiers.begin(), newModifiers.end());
      if (currentModifiers.size() != size)
        result |= ChangeResult::Change;
```
Avoid iterating over the structs twice.

Reply:
done